transport-msg.tex

```diff
@@ -437,3 +429,249 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
+\item An endpoint can be connected to several Message Busses.
+\end{itemize}
+
+A Message Bus implementation must provide interfaces to transmit and receive
+messages with the other side.
+
+While the Bus implementation and how communication of the messages is done
+between the devices and the drivers is completely out of the scope of the
+specification, this chapter provides a set of standard Bus messages to help
+standardizing as much as possible a part of the bus implementation.
+
+\subsubsection{Discovery}\label{sec:Virtio Transport Options / Virtio Over Messages / Message Bus / Discovery}
+
+The Message Bus is responsible of listing the available busses and to enumerate
+all devices available on each bus.
+
+This can be done in a static way, using a device tree for example, or in a way
+using messages to discover from the frontend side which device IDs are
+available on the backend side (using the BUS_MSG_GET_DEVICES bus message).
+
+\subsubsection{Message Format}\label{sec:Virtio Transport Options / Virtio Over Messages / Message Bus / Message Format}
+
+The Bus messages are transfered using the \emph{bus-msg} type in
+VIRTIO_MSG_TYPE and have the following format:
+
+\begin{tabularx}{\textwidth}{|l|l|l|X|}
+\hline
+Name & Offset & Size (bytes) & Content \\
+\hline \hline
+VIRTIO_MSG_TYPE & 0 & 1 & Bit[0]: 0=Request, 1=Answer \newline Bit[1]: 1=bus-msg \newline Bit[2-7]: Reserved \\
+\hline
+VIRTIO_MSG_ID & 1 & 1 & Message ID \\
+\hline
+VIRTIO_MSG_PAYLOAD & 2 & 38 & Payload \\
+\hline
+\end{tabularx}
+
+\subsubsection{Message Definition}\label{sec:Virtio Transport Options / Virtio Over Messages / Message Bus / Message Definition}
+
+The following table is listing the different Virtio-MSG Bus messages and the
+sender for each of them.
+
+\begin{tabular}{|l|l|l|}
+\hline
+Name & ID & Sender \\
+\hline
+\hline
+Reserved                    & 0x0  &        \\
+\hline
+BUS_MSG_ERROR               & 0x1  & Any    \\
+\hline
+BUS_MSG_GET_DEVICES         & 0x2  & Driver \\
+\hline
+BUS_MSG_DEVICE_ADDED        & 0x3  & Device \\
+\hline
+BUS_MSG_DEVICE_REMOVED      & 0x4  & Device \\
+\hline
+BUS_MSG_PING                & 0x5  & Any    \\
+\hline
+BUS_MSG_RESET               & 0x6  & Driver  \\
+\hline
+\end{tabular}
+
+Bus message ID 6-127 are reserved for future use in the specification and
+message ID 128-255 can be used for implementation specific needs.
+
+The following sections are giving more details of the usage of each message
+and the encoding of the payload for the request and answer (when applicable).
+
+\newcommand{\busdef}[1]{\paragraph{BUS_MSG_#1}\label{sec:Virtio Transport Options / Virtio Over Messages / Message Bus / Message Definition / BUS_MSG_#1}}
+
+\busdef{ERROR}
+
+The Error message is used by one side of the bus to inform the other side that
+one message could not be processed correctly. The message contains an error
+code defined using the Standard Posix Error Codes (TODO: add a reference).
```

Comment:
POSIX does not define the numeric value of the errors that it defines (other than reserving zero):

https://pubs.opengroup.org/onlinepubs/9799919799/functions/V2_chap02.html#tag_16_03

Specific implementations pick the numeric values used by different errors on their platform and it's quite messy in practice. Since this protocol will be used to communicate between systems that are running different platforms they may not use the same numeric values for errors. This specification needs to define the actual numeric values for errors that are valid in this protocol.

Reply:
That is a very good point.
We need to come up with a list of errors and define them in the spec instead of the current system pointing to POSIX.
I do not think we need the full POSIX list of errors so we need to come with a reduce subset, at the end there is always the string or IMPDEF solution if an implementation wants to provide more information.
Anybody has something we could start from as a list of error codes ?

Reply:
I noticed that another part of the draft (from admin.tex) defines a set of numeric error codes and says that the values were picked to match what Linux uses for similar errors:

https://github.com/Linaro/virtio-msg-spec/blob/c992db3e403989ae7253b6e2312339f193ccbdef/admin.tex#L174

Perhaps that's a model we could follow? I don't know where that text is from or where it was discussed. It'd be good to have a generic for errors across virtio.

Reply:
Yes it would be a good idea to have some Virtio error code defined in a generic chapter and have admin and virtio message refer to it.
@stsquad Any advice here on how we should propose that ?
We could make an independent patch creating virtio Errors and removing admin ones to refer to it so that we could also refer to it in our patches. Should it be part of our RFC serie ?